### PR TITLE
persist all messages under "guest" (username) and "bestcohort" (channel)

### PR DESCRIPTION
**Pull request by rorysaur**
- load existing messages on mounting the channel component
- persist all new messages
- rerender when any guest sends a new message
- addresses #30

### CODE REVIEW

File: js/actions/actions.js [hunk 1]
```diff
@@ -1,19 +1,69 @@
 var AppDispatcher = require('../dispatcher/app_dispatcher');
 var ActionTypes = require('../constants/constants').ActionTypes;
+var QuackData = require('../data/data');
 var UserStore = require('../stores/user_store');
 var UserCommandHandler = require('../utils/user_command_handler');
 
+var dispatch = function(type, data) {
```
Comment (jacknoble) [inline js/actions/actions.js:7]: me gusta

File: js/actions/actions.js [hunk 2]
```diff
@@ -21,46 +71,44 @@ module.exports = {
...
 
   renameLocalUser: function(newName) {
-    AppDispatcher.dispatch({
-      type: ActionTypes.RENAME_LOCAL_USER,
-      data: {
+    dispatch(
+      ActionTypes.RENAME_LOCAL_USER,
+      {
         newName: newName
       }
```
Comment (jacknoble) [inline js/actions/actions.js]: lets puts the brackets on the same line here

File: js/actions/actions.js [hunk 3]
```diff
@@ -21,46 +71,44 @@ module.exports = {
...
 
   flashNotify: function(message) {
-    AppDispatcher.dispatch({
-      type: ActionTypes.NOTIFY,
-      data: {
+    dispatch(
+      ActionTypes.NOTIFY,
+      {
         message: message
       }
```
Comment (jacknoble) [inline js/actions/actions.js]: same for all the one key payloads

File: js/data/data.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
+var Firebase = require('firebase');
+var FirebaseUtils = require('../utils/firebase');
+
+// TODO put the uri in a constant / env var
+var FirebaseRef = new Firebase('https://quack.firebaseio.com');
+
+module.exports = {
+  create: function(resourceType, options) {
```
Comment (jacknoble) [inline js/data/data.js:8]: Is 'options' the right name for this variable? Seems like "data" or "payload" makes more sense.

Comment (rorysaur) [inline js/data/data.js:8]: it includes success and error callbacks which are not data..

File: js/data/data.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    var handlers = {
+      message: function() {
+        if (!options || !options.message) {
```
Comment (jacknoble) [inline js/data/data.js]: Log an error and return out if no options were specified and then you don't have to keep checking if options were provided.

Comment (rorysaur) [inline js/data/data.js]: kay

Comment (jacknoble) [inline js/data/data.js]: It also seems like callbacks should _always_ be provided and the function shouldn't try to push to firebase if they're not.

File: js/data/data.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    };
+
+    if (handlers[resourceType]) {
+      handlers[resourceType]();
+    }
+  },
+
+  get: function(resourceType, options) {
+    var handlers = {
+      messages: function() {
```
Comment (jacknoble) [inline js/data/data.js:56]: Same here.

File: js/stores/user_store.js [hunk 7]
```diff
@@ -14,7 +14,7 @@ var _users = [
   },
   {
     id: 3,
-    name: 'jack'
+    name: 'guest'
```
Comment (jacknoble) [inline js/stores/user_store.js:17]: :-1:

File: package.json [hunk 8]
```diff
@@ -18,16 +18,16 @@
...
-    "sassify": "^0.7.0"
+    "sassify": "^0.7.0",
+    "watchify": "^2.4.0"
   },
   "browserify": {
     "transform": [
-      [
-        "reactify"
-      ],
+      ["browserify-shim"],
```
Comment (jacknoble) [inline package.json:29]: whats this?

File: js/data/data.js (hunk 6 above)
Comment (jacknoble) [inline js/data/data.js:56]: Actually, it looks like it shouldn't go in any handler, but before the handler is called.

File: package.json (hunk 8 above)
Comment (rorysaur) [inline package.json:29]: browserify-shim is for making js globals require-able (among other capabilities).. we prob wouldn't need it if using ReactFire, but we don't want that so we have to go with their plain js global library

Comment (rorysaur) [inline package.json:29]: their npm client is only for node, not for the browser.. as i understand

**Comment by rorysaur** [comment]
- fixed `{ one-line object style }`
- check for existence of `options` first in every QuackData method
- require success callback in some QuackData methods, i.e. when using "get" or "on", since there is no point to calling it if you're not going to do something with it. for something like "create", i feel it should be optional because you can still accomplish something (persisting new data) without any callbacks. for error callbacks, we COULD require them because it's best practice, but that might be kind of overbearing?

**Comment by jacknoble** [comment]
Sounds good. :ship:  :it:

**Pull request merged**